src/lib.rs

```diff
@@ -342,15 +343,43 @@ where
             continue;
         }
         visited.insert(start);
-        let mut dfs = Dfs::new(graph, start);
-        let mut parent: NodeIndex = start;
-        while let Some(nx) = dfs.next(graph) {
-            if visited.contains(&nx) {
-                continue;
+        let mut children: Vec<NodeIndex> = graph.neighbors(start).collect();
+        children.reverse();
+        let mut stack: Vec<(NodeIndex, Vec<NodeIndex>)> =
+            vec![(start, children)];
+        // Used to track the last position in children vec across iterations
+        let mut index_map: HashMap<NodeIndex, usize> = HashMap::new();
+        index_map.insert(start, 0);
+        while !stack.is_empty() {
+            let temp_parent = stack.last().unwrap();
+            let parent = temp_parent.0;
+            let children = temp_parent.1.clone();
+            let count = *index_map.get(&parent).unwrap();
+            let mut found = false;
+            let mut index = 0;
+            for child in &children {
+                // Skip children for a parent already inspected.
+                if index < count {
+                    index += 1;
+                    continue;
+                }
+                index += 1;
+                if !visited.contains(&child) {
+                    out_vec.push((parent.index(), child.index()));
+                    visited.insert(*child);
+                    let mut grandchildren: Vec<NodeIndex> =
+                        graph.neighbors(*child).collect();
+                    grandchildren.reverse();
+                    stack.push((*child, grandchildren));
+                    index_map.insert(*child, 0);
+                    *index_map.get_mut(&parent).unwrap() = index;
```

Comment:
Does this update the value of `index_map` keyed on `parent`?

Reply:
Yeah, that's what this does. This is basically saying get a mutable reference from the `index_map` HashMap for the key `parent`, unwrap it (ie assuming it's always present) then dereference it (as mutable) and set it's value to index.